java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -152,32 +152,69 @@ private void registerMinion(String minionId, boolean isSaltSSH,
                 return false;
             },
             machineId -> {
+                boolean isReactivation = false;
+                ActivationKey ak = null;
+                ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
+                Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
+                if (managmentKeyLabel.isPresent()) {
+                    ak = ActivationKeyFactory.lookupByKey(managmentKeyLabel.get());
+                    if (ak.getKickstartSession() != null) {
+                        ak.getKickstartSession().markComplete("Installation completed.");
+                    }
+                    if (ak.getServer() == null) {
+                        LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
+                    }
+                    else {
+                        isReactivation = true;
+                    }
+                }
+
                 Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
-                if (checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH)) {
+                if (!isReactivation &&
+                        checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH, grains)) {
                     return true;
                 }
                 // Check if this minion id already exists
-                if (duplicateMinionNamePresent(minionId)) {
+                if (!isReactivation && duplicateMinionNamePresent(minionId)) {
                     return false;
                 }
-                finalizeMinionRegistration(minionId, machineId, creator, saltSSHProxyId, actKeyOverride, isSaltSSH);
+                if (isReactivation) {
+                    reactivateSystem(minionId, machineId, creator, grains, Optional.ofNullable(ak));
+                }
+                finalizeMinionRegistration(minionId, machineId, creator, saltSSHProxyId,
+                        actKeyOverride.isPresent() ? actKeyOverride : Optional.empty(),
+                        isSaltSSH, grains);
                 return true;
             }
         );
     }
 
+    private void reactivateSystem(String minionId, String machineId, Optional<User> creator,
+            ValueMap grains, Optional<ActivationKey> reactivationKey) {
+        // The machine id may have changed, but we know from the reactivation key
+        // which system should become this one
+        reactivationKey.ifPresent(rak -> {
+            Optional<MinionServer> opt = rak.getServer().asMinionServer();
```

Comment:
nested `ifPresent` can be replaced with `flatMap`.

Reply:
@lucidd - I made a change, but please have a look if I did it right.